clientv3/balancer.go

```diff
@@ -354,3 +369,66 @@ func getHost(ep string) string {
 	}
 	return url.Host
 }
+
+func (b *simpleBalancer) trySwitchEndpoint() (doSwitch bool) {
+	var prevPinAddr string
+	var prevEp string
+	doSwitch = true
+
+	func() {
+		b.mu.Lock()
+		defer b.mu.Unlock()
+
+		if len(b.host2ep) <= 1 {
+			doSwitch = false
+			return
+		}
+
+		// in another switching
+		if b.lastPinAddr == b.pinAddr {
+			doSwitch = false
+			return
+		}
+		b.lastPinAddr = b.pinAddr
+		prevPinAddr = b.pinAddr
+		prevEp = b.host2ep[b.pinAddr]
+		delete(b.host2ep, prevPinAddr)
+
+		addrs := make([]grpc.Address, 0, len(b.addrs))
+
+		for _, addr := range b.addrs {
+			if prevPinAddr != addr.Addr {
+				addrs = append(addrs, addr)
+			}
+		}
+		b.addrs = addrs
+
+		select {
+		case b.updateAddrsC <- struct{}{}:
+		case <-b.stopc:
+		}
+	}()
+
+	if !doSwitch {
+		return
+	}
+
+	go func() {
+		b.mu.Lock()
+		for prevPinAddr == b.pinAddr || b.pinAddr == "" {
+			b.pinAddrCond.Wait()
+		}
+
+		b.host2ep[prevPinAddr] = prevEp
```

Comment:
This could corrupt the balancer endpoint list.

1. start trySwitchEndpoint, delete prevPinAddr "a" from the addr list
2. some other goroutine calls client.SetEndpoints("b", "c")
3. trySwitchEndpoint gets notification from pinAddrCond, puts prevEp back into endpoint list.
4. client.Endpoints() == {"a", "b", "c"}

Reply:
OK, I think this is the main reason that this implementation not good. A better way, I think, is not to modify the balancer's endpoints list, but erase the pinAddr, and then put the (endpoints - pinAddr) directly to notifyCh, wait grpc to Up() one of them. it this OK?